miso-web/src/main/webapp/scripts/form.js

```diff
@@ -0,0 +1,249 @@
+   *   getSections: required function(config) returning array of FormSections; see below
+   *   onLoad: optional function(); called after the form is initialized
+   * }
+   * 
+   * FormSection Structure: {
+   *   include: optional boolean; determines whether the section should be displayed. Section is displayed by default
+   *   title: required string; Section heading
+   *   getFields: required function(config) returning array of Fields. See below
+   * }
+   * 
+   * Field Structure: {
+   *   title: string; required except for hidden fields; Field name to display
+   *   data: required string; JSON property to use for value (unless getValue is supplied). Also used as the input ID
+   *   getValue: function(object) returning string; generate a value to display instead of the object's basic field
+   *   type: required string (read-only|hidden|text|dropdown|special); type of field
+   *   include: optional boolean; determines whether the field is displayed. Field is displayed by default
+   *   initial: optional string; value to initialize field value to for new items
+   *   required: optional boolean; whether the field is required
+   *   maxLength: optional integer; maximum number of characters for text input
+   *   getSource: function() returning array of objects; required for dropdown fields; Provides dropdown options
+   *   getItemLabel: function(item) returning string; required for dropdown fields; get the label for a dropdown option
+   *   getItemValue: function(item) returning string; required for dropdown fields; get the value for a dropdown option
+   *   makeControls: function() returning single or array of jQuery controls; required for special fields; set up special fields
+   * }
+   */
+  
+  return {
+    createForm: function(containerId, saveId, object, targetName, config) {
+      var container = $('#' + containerId);
+      var target = FormTarget[targetName];
+      if (!object) {
+        object = {};
+      }
+      
+      writeGeneralValidationBox(container);
+      var sections = getFilteredSections(target, config);
+      sections.forEach(function(section) {
+        writeSection(container, section, object);
+      });
+      
+      $('#' + saveId).click(function() {
+        validateAndSave(containerId, object, target, sections);
+      });
+      
+      if (target.onLoad) {
+        target.onLoad();
+      }
+    }
+  };
+  
+  function validateAndSave(containerId, object, target, sections) {
+    var selector = '#' + containerId;
+    Validate.cleanFields(selector);
+    Validate.clearErrors(selector);
+    
+    sections.forEach(function(section) {
+      section.fields.forEach(function(field) {
+        switch (field.type) {
+        case 'read-only': // Should not be updated
+        case 'special': // FormTarget is responsible for managing updates, likely via an additional hidden field
+          break;
+        case 'hidden':
```

Comment:
I changed how this works while fixing the tests. There is no longer a `hidden` field type, but hidden fields are used when a `read-only` type field displays a value (via `getDisplayValue`) that is different than the object's `data` property value. I think this simplifies the `FormTarget` slightly while better facilitating tests as well.

`read-only` fields are both validated and saved now, as the "read-only" now means "the user cannot directly manipulate this value." It can still be changed via JS. e.g. box `storageLocationId` is a `read-only` field that is modified via the "Change Location" `special` field controls